kamon-akka-http-server/src/main/scala/kamon/akka/http/server/instrumentation/ServerInstrumentation.scala

```diff
@@ -0,0 +1,158 @@
+/* ===================================================
+ * Copyright © 2013 the kamon project <http://kamon.io/>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ========================================================== */
+package kamon.akka.http.server.instrumentation
+
+import akka.http.scaladsl.Http.{ IncomingConnection, ServerBinding }
+import akka.http.scaladsl.model.headers.RawHeader
+import akka.http.scaladsl.model.{ HttpResponse, HttpRequest }
+import akka.http.scaladsl.server.RequestContext
+import kamon.Kamon
+import kamon.akka.http.server.{ AkkaHttpServerExtension, AkkaHttpServer, AkkaHttpServerMetrics }
+import kamon.metric.Entity
+import kamon.trace.{ TraceContextAware, Tracer }
+import org.aspectj.lang.ProceedingJoinPoint
+import org.aspectj.lang.annotation._
+
+@Aspect
+class HttpRequestInstrumentation extends ServerInstrumentationUtils {
+
+  @DeclareMixin("akka.http.scaladsl.model.HttpRequest")
+  def mixinTraceContextAwareToRequest: TraceContextAware = TraceContextAware.default
+
+  @Pointcut("execution(akka.http.scaladsl.model.HttpRequest.new(..)) && this(request)")
+  def httpRequestCreation(request: HttpRequest with TraceContextAware): Unit = {}
+
+  @After("httpRequestCreation(request)")
+  def afterHttpRequestCreation(request: HttpRequest with TraceContextAware): Unit = {
+    val akkaHttpServerExtension = Kamon(AkkaHttpServer)
+    val incomingContext = Tracer.currentContext
+
+    if (incomingContext.isEmpty) {
+      val defaultTraceName = akkaHttpServerExtension.generateTraceName(request)
+      val token = request.headers.find(_.name == akkaHttpServerExtension.settings.traceTokenHeaderName).map(_.value)
+
+      val newContext = Kamon.tracer.newContext(defaultTraceName, token)
+      Tracer.setCurrentContext(newContext)
```

Comment:
I'm not sure if this would be a viable approach for creating a TraceContext, given that a HttpRequest might be created both on the server side and in the client side as well. Typically, when a HttpRequest arrives on the server side we should start a new trace but when a new HttpRequest is created on the client side we should rather create a segment, and only if there is a TraceContext available.

Also, we should always try to match a `Tracer.setCurrentContext(...)` with a `Tracer.clearCurrentContext(...)` or not use these variants at all. It is best to use the `Tracer.withContext(...)` form because it will always put whatever context was before the code blocks executes. It might seem unnecessary at some points, but we have experienced weird behaviors where the TraceContext for different requests get mixed due to leaving dirty threads, so, always clean the context! :D